File: packages/extension/shared/protocol.js
export const PROTOCOL_VERSION = 1;
export const NATIVE_HOST_NAME = 'com.chrome_collect.native_host';
export const EXT_MESSAGE_TYPE = 'EXT_REQUEST';

export const METHODS = {
  HELLO: 'hello',
  BOOKMARK_SAVE: 'bookmark.save',
  BOOKMARK_EXISTS: 'bookmark.existsByUrl',
  BOOKMARK_LIST: 'bookmark.list',
  BOOKMARK_LIST_RECENT: 'bookmark.listRecent',
  BOOKMARK_DELETE: 'bookmark.delete',
  BOOKMARK_GET: 'bookmark.get',
  BOOKMARK_GET_HTML: 'bookmark.getHtml',
  BOOKMARK_UPDATE_ALIAS: 'bookmark.updateAlias',
  BOOKMARK_UPDATE_NOTES: 'bookmark.updateNotes',
  BOOKMARK_DOWNLOAD_HTML: 'bookmark.downloadHtml',
  BOOKMARK_OPEN_FOLDER: 'bookmark.openFolder',
  TRASH_LIST: 'trash.list',
  TRASH_RESTORE: 'trash.restore',
  TRASH_DELETE: 'trash.delete',
  TRASH_EMPTY: 'trash.empty',
  STATS_GET: 'stats.get',
  SETTINGS_GET: 'settings.get',
  SETTINGS_SET_AUTOSTART: 'settings.setAutoStart',
  VERSION_GET: 'version.get',
  UPDATE_START: 'update.start',
  EXTENSION_PING: 'extension.ping',
  UI_OPEN_MANAGER: 'ui.openManager',
  UI_OPEN_EXPORT: 'ui.openExport',
  SHELL_OPEN_EXTERNAL: 'shell.openExternal',
};

export const DEFAULT_TIMEOUT = 15000;

export function formatNativeError(message) {
  if (!message) return 'Native host 错误';
  if (typeof message === 'string') return message;
  if (message.error) return message.error;
  return 'Native host 返回错误';
}

File: packages/tray/internal/protocol/protocol.go
package protocol

import "encoding/json"

const (
	ProtocolVersion = 1
	NativeHostName  = "com.chrome_collect.native_host"
)

const (
	MethodHello               = "hello"
	MethodAppOpenManager      = "ui.openManager"
	MethodShellOpenExternal   = "shell.openExternal"
	MethodBookmarkSave        = "bookmark.save"
	MethodBookmarkExistsByURL = "bookmark.existsByUrl"
	MethodBookmarkList        = "bookmark.list"
	MethodBookmarkListRecent  = "bookmark.listRecent"
	MethodBookmarkGet         = "bookmark.get"
	MethodBookmarkGetHTML     = "bookmark.getHtml"
	MethodBookmarkDelete      = "bookmark.delete"
	MethodBookmarkUpdateAlias = "bookmark.updateAlias"
	MethodBookmarkUpdateNotes = "bookmark.updateNotes"
	MethodBookmarkDownload    = "bookmark.downloadHtml"
	MethodBookmarkOpenFolder  = "bookmark.openFolder"
	MethodTrashList           = "trash.list"
	MethodTrashRestore        = "trash.restore"
	MethodTrashDelete         = "trash.delete"
	MethodTrashEmpty          = "trash.empty"
	MethodStatsGet            = "stats.get"
	MethodSettingsGet         = "settings.get"
	MethodSettingsSetAuto     = "settings.setAutoStart"
	MethodVersionGet          = "version.get"
	MethodUpdateStart         = "update.start"
	MethodExtensionPing       = "extension.ping"
)

type Request struct {
	ID              string          `json:"id"`
	ProtocolVersion int             `json:"protocolVersion"`
	Method          string          `json:"method"`
	Payload         json.RawMessage `json:"payload"`
}

type Response struct {
	ID              string `json:"id"`
	ProtocolVersion int    `json:"protocolVersion"`
	OK              bool   `json:"ok"`
	Result          any    `json:"result,omitempty"`
	Error           *Error `json:"error,omitempty"`
}

type Error struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

func NewSuccess(id string, result any) Response {
	return Response{
		ID:              id,
		ProtocolVersion: ProtocolVersion,
		OK:              true,
		Result:          result,
	}
}

func NewError(id, code, message string) Response {
	return Response{
		ID:              id,
		ProtocolVersion: ProtocolVersion,
		OK:              false,
		Error: &Error{
			Code:    code,
			Message: message,
		},
	}
}
